components/ca-export.js

```diff
@@ -0,0 +1,319 @@
+         * Sets file name
+         * @param {string} value - filename setting
+         */
+        set filename(value) {
+            this.setAttribute('filename', value);
+        }
+
+        /**
+         * Gets cached schema
+         * @returns {object} schema
+         */
+        get schema() {
+            return this._schema;
+        }
+
+        /**
+         * Sets the cache schema
+         * @param {object} value - schema as object
+         */
+        set schema(value = {}) {
+            this._schema = value;
+        }
+
+        /**
+         * Gets cache data
+         * @returns {array} data to export.
+         */
+        get data() {
+            return this._data || [];
+        }
+
+        /**
+         * Gets cached schema
+         * @param {array} value, array of data to export.
+         */
+        set data(value) {
+            this._data = value;
+
+            if (this._data && this._data.length > 0) {
+                this.hasData = true;
+            } else {
+                this.hasData = false;
+            }
+
+            if (this.showRecordCount) {
+                this.button.textContent += ` (${(this.data || []).length} records)`;
+            }
+        }
+
+        /**
+         * Gets has data attribute
+         * @returns {boolean} setting for has data status
+         */
+        get hasData() {
+            return this.getAttribute('has-data');
+        }
+
+        /**
+         * Sets hasData vflag
+         * @param {boolean} value, boolean to trigger has data/
+         */
+        set hasData(value) {
+            this.setAttribute('has-data', value === true);
+        }
+
+        /**
+         * Gets export type
+         * @returns {string} export setting
+         */
+        get type() {
+            return this.getAttribute('type') || 'csv';
+        }
+
+        /**
+         * Sets the type of export
+         * @param {string} value, sets the export type attribute.
+         */
+        set type(value) {
+            // ensure its always csv or xls
+            const fileType = value !== 'xlsx' ? 'csv' : 'xlsx';
+            this.setAttribute('type', fileType);
+        }
+
+        /**
+         * Gets the template converstion url
+         * @returns {string} url for conversion
+         */
+        get templateConversionUrl() {
+            return this.getAttribute('template-conversion-url') || '/excel-download';
+        }
+
+        /**
+         * Sets the type of export
+         * @param {string} value, conversion url, default to /excel-download
+         */
+        set templateConversionUrl(value = '/excel-download') {
+            this.setAttribute('template-conversion-url', value);
+            this.form.action = value;
+        }
+
+       /**
+        * onbeforedownload - Called before download, returning false cancels the download
+        * @param {Event} e, event object.
+        * @returns {undefined} void function.
+        */
+        onbeforedownload() {
```

Comment:
we check for the existence of both these methods, shouldn't we just refactor these out?